### ISSUE TITLE
[clangd] feature request: Add ast matcher API to clangd

I would love to have a feature in clangd, which enables editor plugins to extend the search capabilities of that editor with an AST search. 

### ISSUE THREAD

**Issue by Febbe** [llvm/llvm-project#66365]
While refactoring or reviewing code I often find problematic code, which occurs in similar forms in the complete codebase. Searching for those via a regex ist cumbersome and does not work for all cases.

My preferred workflow would be:
1. Activate AST matching
1. Select code
2. Press Ctrl+F / Ctrl+H to retrieve the current AST in the search field
3. Edit Ast to be more generic
4. Replace AST with other AST (In case of replacement)

In order to implement this in clangd-plugins, clangd itself requires the functionality to 
1. Gain the AST from a code selection
2. Gain a view of sourcecode-location-range objects from an AST
3. Allow AST replacements for a sourcecode-location-range object.

**Comment by torshepherd** [comment]
So clangd does have support for showing AST, just select some text and right click

**Comment by Febbe** [comment]
No, not showing the AST, I know, that clangd can print it. I want to search and replace in all ASTs of all files with the help of the Clang AST Matcher DSL, just like I can do it either via clang-query or writing my own libtooling program.
As I know clang query only supports single files and does not allow replacements and writing a libtooling parser is most often overkill, since I will never need that parser again.
